### ISSUE TITLE
Server can't open stream to client on an active connection

### ISSUE THREAD

**Issue by Wondertan** [quic-go/quic-go#4712]
## Problem Overview
I am running a go-libp2p network with quic-go@v0.47.0 in k8s cluster with 3 servers and 5 clients deployed as pods. Each client connects to each server over QUIC transport and successfully initiates multiple streams. Servers similarly perform well initially. They respond to new streams, open new streams to clients, and maintain long-standing streams(like GossipSub ones).

However, in  ~10-30 minutes, servers lose the ability to open new streams to clients—every attempt to open a stream from server to client times out by context. Clients can freely open streams and receive data over long-standing streams, but servers cannot. Pinging a server from a client using libp2p's ping protocol is instantaneous, while pinging the client from the server never succeeds even though they are on the same connection as reported by libp2p. 

### Why not libp2p issue?
* TCP/WRTC transport work well(although WebTransport similarly fails)
* Rcmgr's, go-multistream, swarm behave correctly according to logs and metrics.

## Other
* I am exploring QLOG and will return with more info if I find anything there. So far looking at qlogs didn't yield anything concrete, so some tips for noobies are appreciated. 
* My current hypothesis is that something needs to be fixed with the infra setup for UDP, as this is not reproducible on WAN. However, our DevOps team and I are unsure what could go wrong there, so any external experience running QUIC in a K8s environment would help. Also, the fact that the issue is strictly unidirectional (Server -> Client) makes me wonder if that's a software bug.

**Comment by marten-seemann** [comment]
I'd be surprised if this was a UDP issue. If UDP was blocked, the connection would time out.

QUIC enforces a maximum number of concurrent streams (`Config.MaxIncomingStreams`), and `Connection.OpenStreamSync` will block until new stream limit is granted by the peer.

Can you check how many streams both peers believe to be opened when this situation occurs?

**Comment by Wondertan** [comment]
👍🏻
> I'd be surprised if this was a UDP issue. If UDP was blocked, the connection would time out.

According to rcmgr metrics, all clients have never exceeded `25` streams, while `MaxIncomingStreams` in libp2p defaults to `256`
> Can you check how many streams both peers believe to be opened when this situation occurs?

**Comment by marten-seemann** [comment]
There might be a mismatch between libp2p and quic-go, i.e. quic-go might consider a stream open that libp2p considers closed (of course, this would be a bug!).

Unfortunately, quic-go's Prometheus metrics are quite rudimentary at the moment, and don't cover streams yet, but maybe you can add some debug statements in the streams map?

**Comment by Wondertan** [comment]
Do you think QLOG could help here? 

Do you refer to streams map inside of QUIC connection?

**Comment by marten-seemann** [comment]
> Do you refer to the streams map inside of QUIC connection in quic-go's codebase?

Yes: https://github.com/quic-go/quic-go/blob/master/streams_map.go

> Do you think QLOG could help here?

Maybe, but we might need to track the state of at least 256 streams, which might be a bit cumbersome to do by hand.

**Comment by Wondertan** [comment]
Ok, I am still getting familiar with this codebase. 

So your hypothesis is that libp2p closes streams and considers them closed, while quic-go still has them lurking around. I got it.

Where do you recommend injecting those debug statements? On the client or server-side? Also, can we use debug logged for it?

**Comment by Wondertan** [comment]
Gentle ping @marten-seemann

**Comment by Wondertan** [comment]
Ok, I've added logs in https://github.com/Wondertan/quic-go/commit/b43db319a92055e3f31cd122355b3d52c4ba5454.

<details><summary>Early log bunch</summary>
<p>
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11681, active streams 38
2024/10/30 03:21:38 incoming_streams delete stream 11464, active streams 39
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11682, active streams 38
2024/10/30 03:21:38 incoming_streams delete stream 11465, active streams 39
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11683, active streams 38
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11685, active streams 39
2024/10/30 03:21:38 incoming_streams delete stream 11468, active streams 39
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11686, active streams 38
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11688, active streams 39
2024/10/30 03:21:38 incoming_streams delete stream 11471, active streams 39
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11689, active streams 38
2024/10/30 03:21:38 incoming_streams delete stream 11472, active streams 39
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11690, active streams 38
2024/10/30 03:21:38 incoming_streams delete stream 11473, active streams 39
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11691, active streams 38
2024/10/30 03:21:38 incoming_streams delete stream 11474, active streams 39
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11692, active streams 38
2024/10/30 03:21:38 incoming_streams delete stream 11475, active streams 39
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11693, active streams 38
2024/10/30 03:21:38 incoming_streams delete stream 11476, active streams 39
2024/10/30 03:21:38 incoming_streams sent MAX_STREAM_ID 11694, active streams 38
</p>
</details> 

<details><summary>Ending logs, right before the deadlock</summary>
<p>
2024/10/30 03:32:35 incoming_streams delete stream 58930, active streams 255
2024/10/30 03:32:35 incoming_streams sent MAX_STREAM_ID 58932, active streams 254
2024/10/30 03:32:35 incoming_streams delete stream 58932, active streams 256
2024/10/30 03:32:35 incoming_streams sent MAX_STREAM_ID 58933, active streams 255
2024/10/30 03:32:35 incoming_streams delete stream 58933, active streams 256
2024/10/30 03:32:35 incoming_streams sent MAX_STREAM_ID 58934, active streams 255
2024/10/30 03:32:35 incoming_streams delete stream 58931, active streams 255
2024/10/30 03:32:35 incoming_streams sent MAX_STREAM_ID 58935, active streams 254
2024/10/30 03:32:35 incoming_streams delete stream 58934, active streams 256
2024/10/30 03:32:35 incoming_streams sent MAX_STREAM_ID 58936, active streams 255
2024/10/30 03:32:35 incoming_streams delete stream 58935, active streams 256
2024/10/30 03:32:35 incoming_streams sent MAX_STREAM_ID 58937, active streams 255
2024/10/30 03:32:35 incoming_streams delete stream 58936, active streams 255
2024/10/30 03:32:35 incoming_streams sent MAX_STREAM_ID 58938, active streams 254
2024/10/30 03:32:35 incoming_streams delete stream 58937, active streams 256
2024/10/30 03:32:35 incoming_streams sent MAX_STREAM_ID 58939, active streams 255
2024/10/30 03:32:35 incoming_streams delete stream 58938, active streams 255
2024/10/30 03:32:35 incoming_streams sent MAX_STREAM_ID 58940, active streams 254
2024/10/30 03:32:35 incoming_streams delete stream 58939, active streams 255
2024/10/30 03:32:35 incoming_streams sent MAX_STREAM_ID 58941, active streams 254
2024/10/30 03:32:35 incoming_streams delete stream 58940, active streams 255
2024/10/30 03:32:35 incoming_streams sent MAX_STREAM_ID 58942, active streams 254
</p>
</details> 

Active streams are clearly leaking. What's unclear to me is how the number of streams is incremented right after being decremented:
* deletes stream 58938 -> active streams goes from 255 -> 254
* deletes stream 58939 -> active streams are back to 255 and again decremented to 254

**Comment by marten-seemann** [comment]
Sorry for the late reply, quite busy right now.



> Active streams are clearly leaking. What's unclear to me is how the number of streams is incremented right after being decremented:
> 
> * deletes stream 58938 -> active streams goes from 255 -> 254
> * deletes stream 58939 -> active streams are back to 255 and again decremented to 254

Here's what I suspect is happening:
* you send MAX_STREAMS frame, granting new stream credit to the peer
* the peer immediately uses this credit to open a new stream

> According to rcmgr metrics, all clients have never exceeded the total of `15` streams, while `MaxIncomingStreams` in libp2p defaults to `256` and is per peer.

Can you double-check this with the swarm metrics as well? They _should_ coincide, but if they don't, this could point us to a bug in libp2p's swarm implementation.

**Comment by Wondertan** [comment]
Swarm's connection state reports the number equal to rcmgr

**Comment by marten-seemann** [comment]
Yes, this sounds very much like a the problem is in libp2p's QUIC wrapper. We have a number of tests for this logic in quic-go, and I haven't heard any complaints about this happening outside of libp2p.

**Comment by marten-seemann** [comment]
Yes, this sounds very much like a the problem is in libp2p's QUIC wrapper. We have a number of tests for this logic in quic-go, and I haven't heard any complaints about this happening outside of libp2p.

cc @sukunrt @MarcoPolo

**Comment by Wondertan** [comment]
> Here's what I suspect is happening:
> *  you send MAX_STREAMS frame, granting new stream credit to the peer
> *  the peer immediately uses this credit to open a new stream

This makes sense, and this is the correct behavior we would expect. 

There are also logs that confirm streams are not deleted, which proves that the stream initiator doesn't cancel stream under certain condition.

**Comment by Wondertan** [comment]
Examining the QUIC wrapper in libp2p, I don't see anything suspicious, as the logic there is very straightforward. One thing that stands out is calling both `CancelRead` and then `Close` in the [Stream.Close](https://github.com/libp2p/go-libp2p/blob/362e5836f1468aa7f158d5da157c0e99da334d06/p2p/transport/quic/stream.go#L44-L45). I am not sure why `CancelRead` has to be called before `Close` and I can see examples in quic only calling `Close`. I am gonna try to reproduce without `CancelRead` now.

**Comment by marten-seemann** [comment]
Yes, there's a mismatch between the stream APIs that's proved painful in the past. The reason is that libp2p doesn't embrace the unidirectionality of the send and receive side of the stream. It's not clear what `Reset` actually means if you think of a stream as (conceptually) composed of two unidirectional streams in opposite direction, so they have to jump through some hoops. Not sure if that's what the issue is though.

To debug this issue, it might be interesting to identify a particular stream which libp2p considers closed, but quic-go doesn't, and work backwards from there. To do that, here's what you could do: Log the stream state transitions inside of libp2p (together with the QUIC stream ID), and record a qlog (no need to add logging in quic-go, it's all in the qlog). You can get the stream ID from the `StreamID` method on the stream via a type assertion.

**Comment by Wondertan** [comment]
So, what's the correct way to close QUIC streams? Only `Close` or also `CancelRead`? I don't get what you mean by not embracing "unidirectionality." It makes it sound like both streams should be closed independently.

I tried to remove `CancelRead` from `Close` and almost started celebrating because I couldn't reproduce the original behavior. However..., this only pushed the problem elsewhere. Instead of the server getting stuck consistently, now the clients get stuck consistently, and surprisingly, the server works well. I followed debugging this path to ensure `Close` is called once only and nothing else besides `Close`, but that keeps the situation the same (FYI https://github.com/Wondertan/go-libp2p/commit/dfbac09cdc3d4a7bd6d64787bb0886d06976d6c8)

**Comment by Wondertan** [comment]
By the way, I am almost certain it's not a libp2p issue. The [SwarmStream.Close](https://github.com/libp2p/go-libp2p/blob/362e5836f1468aa7f158d5da157c0e99da334d06/p2p/net/swarm/swarm_stream.go#L80-L84) calls the underlying stream close first and then proceeds with logic related to the swarm. And we know that the swarm's logic behaves correctly there and doesn't block. Thus, I am skeptical that going the path of debugging libp2p's state vs quic would be fruitful.

**Comment by MarcoPolo** [comment]
Any chance you have a repro I could try? I'm suspicious of the discrepancy of number of streams. It very well could be some tricky race in the swarm.

**Comment by Wondertan** [comment]
> Any chance you have a repro I could try?

I wish 😢. I tried to write a test to make that reproducible locally, but with no luck. The feedback loop between adding a change and deploying to the cluster is extremely slow and error-prone. We could try to give you access to our infra, but that's a whole story on its own

**Comment by Wondertan** [comment]
> It very well could be some tricky race in the swarm.

Q: As long as `NewStream` returns a stream that points to the correct QUIC stream, that shouldn't be possible, right? If the swarm's stream calls quic's stream close immediately, is there any possibility of a race?

**Comment by MarcoPolo** [comment]
I admit, I don't fully understand the setup here. But calling close on a stream returned from the swarm does call the close method on the quic transport stream right away. 

Does the problem show up faster if you lower the MaxIncomingStreams [here](https://github.com/libp2p/go-libp2p/blob/362e5836f1468aa7f158d5da157c0e99da334d06/p2p/transport/quicreuse/config.go#L10)?

Does this happen if the server/client are on the same network? Does latency affect how often this comes up? Does either side reset the stream, or do they both close cleanly?

**Comment by sukunrt** [comment]
What do the corresponding logs for https://github.com/quic-go/quic-go/issues/4712#issuecomment-2445794652 on the server side look like? 

What do rcmgr and swarm metrics on the server side look like? Are they also around 15 streams? 

My theory is: 
Client is done with the stream and calls close. This cleans up all libp2p resources, but not quic-go ones. quic-go needs to keep the stream around and re transmit all the data till the server reads it. 

Do you know who closes the stream first? The server(initiator of the stream in our case) or the client?

**Comment by marten-seemann** [comment]
> My theory is:
> Client is done with the stream and calls close. This cleans up all libp2p resources, but not quic-go ones. quic-go needs to keep the stream around and re transmit all the data till the server reads it. This explains the mismatch between quic and libp2p stream counts.

@sukunrt This is correct, but I'm not sure if this is actually the issue we're seeing here. Repairing packet loss should be fairly quick (O(few RTTs) at most), assuming that the peer is not malicious and intentionally withholding acknowledgments for these STREAMs from us.

I understand that the feedback loop is annoying here, but I think [my earlier suggestion](https://github.com/quic-go/quic-go/issues/4712#issuecomment-2448928021) is our best chance of debugging what's actually going on. @Wondertan Any chance you could deploy this and post the logs here?

**Comment by Wondertan** [comment]
> Log the stream state transitions inside of libp2p (together with the QUIC stream ID), and record a qlog (no need to add logging in quic-go, it's all in the qlog). 

@marten-seemann, which stream state transitions do you mean exactly? New and Close? 

Then I share a qlog with libp2p logs for you to analyze?

**Comment by Wondertan** [comment]
> What do rcmgr and swarm metrics on the server side look like? Are they also around 15 streams?

@sukunrt, yes, libp2p shows nothing suspicious.

> Do you know who closes the stream first? The server(initiator of the stream in our case) or the client?

The server closes it first. This happens in Bitswap, where the server sends messages in new streams and closes them immediately after the data is Written.

**Comment by Wondertan** [comment]
> Does the problem show up faster if you lower the MaxIncomingStreams [here](https://github.com/libp2p/go-libp2p/blob/362e5836f1468aa7f158d5da157c0e99da334d06/p2p/transport/quicreuse/config.go#L10)?

@MarcoPolo, I haven't tried lowering it, but the new streams get stuck when there are 255-256 active streams as reported per logs, which doesn't wholly confirm connection to the 256 MaxIncomingStreams limit, but the odds of that are very high.


> Does this happen if the server/client are on the same network? 

They are in the same 100.x.x.x private network.

> Does latency affect how often this comes up? 

I haven't seen any latency connection so far. It's all in the same cluster/datacenter with 1ms latencies.

> Does either side reset the stream, or do they both close cleanly?

Cleanly with no errors coming out of `Close`. I checked that. 
I only sometimes see `close called for canceled stream %d` but haven't found correlations with our problem.

**Comment by MarcoPolo** [comment]
Can you try this quick and dirty patch to see if there's a mismatch between what libp2p may consider closed and what quic-go is considering closed? https://gist.github.com/MarcoPolo/9d21a5b3472d37f091237fda31687c52

That tries to track active streams for a connection in the most straightforward way possible. That number should be the same as what libp2p reports. 

I think it should also be the same as what quic-go tracks for its own stream count (with some timing variance, since it could read a reset stream frame before the application is made aware of it). But if it isn't, then it's possible there's an error in the streamCompleted logic in quic-go. 

If the server is blocked in opening new streams to the client (server can't ping client), then the issue sounds like it is on the client side (it may not be accounting properly).

---

> So, what's the correct way to close QUIC streams? Only Close or also CancelRead?

If you want to consider a stream done (you won't right or read anymore) then Close (to close your send side) _and_ CancelRead (to inform the peer that you are not reading anymore).

**Comment by Wondertan** [comment]
I will need sometime try this due to Devcon next week, but will surely get to it

**Comment by marten-seemann** [comment]
Thank you @Wondertan!

**Comment by MarcoPolo** [comment]
Also @Wondertan make sure to include a qlog as well please. Thanks!

**Comment by marten-seemann** [comment]
@Wondertan Do you have any update on this?

**Comment by Wondertan** [comment]
No. I am OOO this week and getting to it the next one.

**Comment by Wondertan** [comment]
Update: I am back at this and could reproduce the issue with the recent versions of both quic and libp2p.  
Next: Rerun with Marco's patch

**Comment by Wondertan** [comment]
__Update__: I reproduced the issue with the Marco's patch. However, I didn't see any logs, which means [the threshold](https://gist.github.com/MarcoPolo/9d21a5b3472d37f091237fda31687c52#file-debug-patch-L18) wasn't reached, which means streams were properly closed on the libp2p side.

__Next__: Before making any conclusions, I am going to verify again that the patch was applied correctly and that the stalling of nodes is coming from QUIC by injecting print statements I did previously.  I will also need to capture QLOG for the nodes because I forgot on the last run.

**Comment by Wondertan** [comment]
**Update:** I can see this is the same issue, and the number of streams is reaching 256 on the QUIC side.
**Next:** Post and look into QLOG. I'm going to do it the next day.

**Comment by marten-seemann** [comment]
3.5 GB is way too large to process in qvis. We'd need something O(100 MB).

**Comment by marten-seemann** [comment]
I think I might have found the culprit: #4803.

@Wondertan Could you deploy https://github.com/quic-go/quic-go/pull/4804 and see if that fixes the issue?

**Comment by Wondertan** [comment]
Really are great news! Will try it once asap

**Comment by Wondertan** [comment]
I tried and I think it is fixed. I've been running the repro today and monitoring it from time to time. It usually took up to 2 hours to deadlock, but this time I got it beyond. For ~5 hours the node was sampling 8mb blocks testnet until it sank again. Seeing it locking again was a big disappointment, but upon further investigation I realized that it could be unrelated to the problem. The bootstrapper restarted in k8s causing address change. I am gonna run this again rn for the overnight to see if the full network syncing completes and then we can call it an issue.

**Comment by marten-seemann** [comment]
Thank you @Wondertan!

**Comment by Wondertan** [comment]
Hey @marten-seemann, I still don't have full confirmation of the issue being fixed. The other reason restarting the bootstrapper is still bugging me and a mystery. k8s does not show any trace of whats happening, so I for now stuck with devops folks helping me. I would like to confirm this definitively, so bear with me

**Comment by marten-seemann** [comment]
Sure, feel free to reopen if you run into any more problems!

**Comment by Wondertan** [comment]
Coming back to confirm that the issue was unrelated and now I can fully sync solely using QUIC 🎉 

I still have a bit of worry about those being STREAM_RESET rather than STOP_SENDING and
why packets are being lost in that environment for loss detection to kick in

**Comment by marten-seemann** [comment]
> I still have a bit of worry about those being STREAM_RESET rather than STOP_SENDING and
> why packets are being lost in that environment for loss detection to kick in

What's your worry here?

**Comment by Wondertan** [comment]
I mean I don't know how normal it is for k8s to loose packets. It seems like in closed and controlled environment like ours that shouldn't happen and there might be something else fishy going on. 

Also, `STREAM_RESET` seem to be a result of application calling `stream.Reset` as opposed to `stream.Close`, but I didn't observe any resets or issues to cause that, besides resets caused by `close called for canceled stream 5952`, although it resetting an already closed stream.

Also, what's the timeline for the new release?
